File: src/main/java/org/example/pidev/model/Event.java
package org.example.pidev.model;

import java.time.LocalDate;

public class Event {
    private int id;
    private String name;
    private String description;
    private LocalDate date;
    private String location;
    private String organiser;
    private String eventType;
    private int nbParticipant;
    private float ticketPrice;
    private boolean hasFormation;
    private int formationId;
    private Formation formation;
    private float longitude;
    private float latitude;

    // Default constructor
    public Event() {}

    // Constructor including longitude and latitude
    public Event(int id, String name, String description, LocalDate date, String location, String organiser, String eventType, int nbParticipant, float ticketPrice, boolean hasFormation, int formationId, Formation formation, float longitude, float latitude) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.date = date;
        this.location = location;
        this.organiser = organiser;
        this.eventType = eventType;
        this.nbParticipant = nbParticipant;
        this.ticketPrice = ticketPrice;
        this.hasFormation = hasFormation;
        this.formationId = formationId;
        this.formation = formation;
        this.longitude = longitude;
        this.latitude = latitude;
    }

    // Constructor without formation
    public Event(int id, String name, String description, LocalDate date, String location, String organiser, String eventType, int nbParticipant, float ticketPrice, boolean hasFormation) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.date = date;
        this.location = location;
        this.organiser = organiser;
        this.eventType = eventType;
        this.nbParticipant = nbParticipant;
        this.ticketPrice = ticketPrice;
        this.hasFormation = hasFormation;
        this.formationId = 0; // Default to no formation
        this.formation = null; // No formation by default
        this.longitude = 0.0f;  // Default longitude
        this.latitude = 0.0f;   // Default latitude
    }

    // Getters and setters for longitude and latitude
    public float getLongitude() {
        return longitude;
    }

    public void setLongitude(float longitude) {
        this.longitude = longitude;
    }

    public float getLatitude() {
        return latitude;
    }

    public void setLatitude(float latitude) {
        this.latitude = latitude;
    }

    // Getters and setters for other fields...
    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public String getName() { return name; }
    public void setName(String name) { this.name = name; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public LocalDate getDate() { return date; }
    public void setDate(LocalDate date) { this.date = date; }

    public String getLocation() { return location; }
    public void setLocation(String location) { this.location = location; }

    public String getOrganiser() { return organiser; }
    public void setOrganiser(String organiser) { this.organiser = organiser; }

    public String getEventType() { return eventType; }
    public void setEventType(String eventType) { this.eventType = eventType; }

    public int getNbParticipant() { return nbParticipant; }
    public void setNbParticipant(int nbParticipant) { this.nbParticipant = nbParticipant; }

    public float getTicketPrice() { return ticketPrice; }
    public void setTicketPrice(float ticketPrice) { this.ticketPrice = ticketPrice; }

    public boolean isHasFormation() { return hasFormation; }
    public void setHasFormation(boolean hasFormation) {
        this.hasFormation = hasFormation;
        if (!hasFormation) {
            this.formationId = 0;
            this.formation = null;
        }
    }

    public int getFormationId() { return formationId; }
    public void setFormationId(int formationId) {
        this.formationId = formationId;
        this.hasFormation = (formationId > 0);
    }

    public Formation getFormation() { return formation; }
    public void setFormation(Formation formation) {
        this.formation = formation;
        if (formation != null) {
            this.formationId = formation.getId();
            this.hasFormation = true;
        }
    }

    @Override
    public String toString() {
        return "Event{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", description='" + description + '\'' +
                ", date=" + date +
                ", location='" + location + '\'' +
                ", organiser='" + organiser + '\'' +
                ", eventType='" + eventType + '\'' +
                ", nbParticipant=" + nbParticipant +
                ", ticketPrice=" + ticketPrice +
                ", hasFormation=" + hasFormation +
                ", formationId=" + formationId +
                ", formation=" + (formation != null ? formation.getTitre() : "No Formation") +
                ", longitude=" + longitude +
                ", latitude=" + latitude +
                '}';
    }
}

File: src/Entity/Event.php
<?php
// src/Entity/Event.php

namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;
use Symfony\Component\Validator\Context\ExecutionContextInterface;

#[ORM\Entity(repositoryClass: "App\Repository\EventRepository")]
#[ORM\Table(name: "events")]
class Event
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: "integer")]
    private ?int $id = null;

    #[ORM\Column(type: "string", length: 255)]
    #[Assert\NotBlank]
    #[Assert\Length(min: 3, max: 255)]
    private string $name;

    #[ORM\Column(type: "text", nullable: true)]
    #[Assert\Length(max: 1000)]
    private ?string $description = null;

    #[ORM\Column(type: "date")]
    #[Assert\NotBlank]
    #[Assert\GreaterThan('today')]
    private \DateTimeInterface $date;

    #[ORM\Column(type: "string", length: 255)]
    #[Assert\NotBlank]
    #[Assert\Length(min: 3, max: 255)]
    private string $location;

    #[ORM\Column(type: "string", length: 255)]
    #[Assert\NotBlank]
    #[Assert\Length(min: 3, max: 255)]
    private string $organiser;

    #[ORM\Column(type: "string", length: 255)]
    #[Assert\NotBlank]
    #[Assert\Choice(choices: ['Conference', 'Workshop', 'Seminar', 'Webinar'], message: 'Choose a valid event type')]
    private string $eventType;

    #[ORM\Column(type: "integer")]
    #[Assert\PositiveOrZero]
    #[Assert\LessThanOrEqual(1000)]
    private int $nbParticipant = 0;

    #[ORM\Column(type: "float")]
    #[Assert\PositiveOrZero]
    #[Assert\LessThanOrEqual(1000)]
    private float $ticketPrice = 0.0;

    #[ORM\Column(type: "boolean")]
    private bool $hasFormation = false;

    #[ORM\Column(type: "integer", nullable: true)]
    #[Assert\IsNull(groups: ["has_formation_false"])]
    private ?int $formationId = null;

    #[ORM\Column(type: "float")]
    #[Assert\Range(min: -180, max: 180)]
    private float $longitude = 0.0;

    #[ORM\Column(type: "float")]
    #[Assert\Range(min: -90, max: 90)]
    private float $latitude = 0.0;

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function setName(string $name): self
    {
        $this->name = $name;
        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(?string $description): self
    {
        $this->description = $description;
        return $this;
    }

    public function getDate(): \DateTimeInterface
    {
        return $this->date;
    }

    public function setDate(\DateTimeInterface $date): self
    {
        $this->date = $date;
        return $this;
    }

    public function getLocation(): string
    {
        return $this->location;
    }

    public function setLocation(string $location): self
    {
        $this->location = $location;
        return $this;
    }

    public function getOrganiser(): string
    {
        return $this->organiser;
    }

    public function setOrganiser(string $organiser): self
    {
        $this->organiser = $organiser;
        return $this;
    }

    public function getEventType(): string
    {
        return $this->eventType;
    }

    public function setEventType(string $eventType): self
    {
        $this->eventType = $eventType;
        return $this;
    }

    public function getNbParticipant(): int
    {
        return $this->nbParticipant;
    }

    public function setNbParticipant(int $nbParticipant): self
    {
        $this->nbParticipant = $nbParticipant;
        return $this;
    }

    public function getTicketPrice(): float
    {
        return $this->ticketPrice;
    }

    public function setTicketPrice(float $ticketPrice): self
    {
        $this->ticketPrice = $ticketPrice;
        return $this;
    }

    public function isHasFormation(): bool
    {
        return $this->hasFormation;
    }

    public function setHasFormation(bool $hasFormation): self
    {
        $this->hasFormation = $hasFormation;
        if (!$hasFormation) {
            $this->formationId = null;
        }
        return $this;
    }

    public function getFormationId(): ?int
    {
        return $this->formationId;
    }

    public function setFormationId(?int $formationId): self
    {
        $this->formationId = $formationId;
        $this->hasFormation = ($formationId !== null);
        return $this;
    }

    public function getLongitude(): float
    {
        return $this->longitude;
    }

    public function setLongitude(float $longitude): self
    {
        $this->longitude = $longitude;
        return $this;
    }

    public function getLatitude(): float
    {
        return $this->latitude;
    }

    public function setLatitude(float $latitude): self
    {
        $this->latitude = $latitude;
        return $this;
    }

    #[Assert\Callback]
    public function validateHasFormationAndFormationId(ExecutionContextInterface $context)
    {
        if (!$this->hasFormation && $this->formationId !== null) {
            $context->buildViolation('Formation ID should be null if hasFormation is false')
                ->addViolation();
        }
    }
    public function __toString(): string
    {
        return sprintf(
            'Event{id=%d, name="%s", description="%s", date=%s, location="%s", organiser="%s", eventType="%s", nbParticipant=%d, ticketPrice=%.2f, hasFormation=%s, formationId=%s, longitude=%.6f, latitude=%.6f}',
            $this->id,
            $this->name,
            $this->description,
            $this->date->format('Y-m-d'),
            $this->location,
            $this->organiser,
            $this->eventType,
            $this->nbParticipant,
            $this->ticketPrice,
            $this->hasFormation ? 'true' : 'false',
            $this->formationId ?? 'null',
            $this->longitude,
            $this->latitude
        );
    }
}